[Excel](data types) New APIs for linked entities (#5091)
OfficeDev/office-js-docs-pr: docs/excel/add-properties-to-basic-cell-values.md

---
title: Add properties to basic cell values
description: Add properties to basic cell values.
ms.topic: how-to #Required; leave this attribute/value as-is
ms.date: 03/17/2025
ms.localizationpriority: medium
---

# Add properties to basic cell values

Add properties to basic cell values in Excel to associate additional information with the values. Similar to entity values, you can add properties to the **string**, **double**, and **Boolean** basic types. Each property is a key/value pair. The following example shows the number 104.67 (double) that represents a bill with added fields named **Drinks**, **Food**, **Tax**, and **Tip**.

:::image type="content" source="../images/data-type-basic-fields.png" alt-text="Screen shot of the drinks, food, tax, and tip fields shown for the selected cell value.":::

> If the user chooses to show the data type card, they will see the values for the fields.

Reviewer: ```suggestion
If the user chooses to show the data type card, they'll see the values for the fields.
```

:::image type="content" source="../images/data-type-basic-data-type-card.png" alt-text="Data type card showing values for drinks, food, tax, and tip properties":::

Cell value properties can also be used in formulas.

> :::image type="content" source="../images/data-type-basic-dot-syntax.png" alt-text="Show user typing "a1." and Excel showing a menu with drinks, food, tax, and tip options.":::

Reviewer 2: I think this will fix the publishing error.
```suggestion
:::image type="content" source="../images/data-type-basic-dot-syntax.png" alt-text="Show user typing 'a1.' and Excel showing a menu with drinks, food, tax, and tip options.":::
```

## Create a cell value with properties

> To create a cell value and add properties to it, use **valuesAsJson** to assign properties. The following objects offer the **valuesAsJson** property.

Author: thanks yeah that's a mistake in the conversion from the word doc.

> To create a cell value and add properties to it, use **valuesAsJson** to assign properties. The following objects offer the **valuesAsJson** property.

Author: fixed.

```javascript
async function createNumberProperties() {
  await Excel.run(async (context) => {
    const sheet = context.workbook.worksheets.getActiveWorksheet();
    const range = sheet.getRange("A1");
    range.valuesAsJson = [
      [
        {
          type: Excel.CellValueType.double,
          basicType: Excel.RangeValueType.double,
          basicValue: 104.67,
          properties: {
            Food: {
              type: Excel.CellValueType.string,
              basicType: Excel.RangeValueType.string,
              basicValue: "Sandwich and fries"
            },
            Drinks: {
              type: Excel.CellValueType.string,
              basicType: Excel.RangeValueType.string,
              basicValue: "Soda"
            },
            Tax: {
              type: Excel.CellValueType.double,
              basicType: Excel.RangeValueType.double,
              basicValue: 5.5
            },
            Tip: {
              type: Excel.CellValueType.double,
              basicType: Excel.RangeValueType.double,
              basicValue: 21
            }
          }
        }
      ]
    ];
    await context.sync();
  });
}
```

> [!NOTE]

> > Some cell values change based on a user's locale. The **valuesAsJsonLocal** property offers localization support and is available on all the same objects as **valuesAsJson**.

Reviewer: Thank you for being inclusive here!

```javascript
// This is an example of the complete JSON of a formatted number value with a property.
// In this case, the number is formatted as currency.
async function createCurrencyValue() {
  await Excel.run(async (context) => {
    const sheet = context.workbook.worksheets.getActiveWorksheet();
    const range = sheet.getRange("A1");
    range.valuesAsJson = [
      [
        {
          type: Excel.CellValueType.double,
          basicType: Excel.RangeValueType.double,
          basicValue: 24,
          numberFormat: "$0.00",
          properties: {
            Name: {
              type: Excel.CellValueType.string,
              basicValue: "dollar"
            }
          }
        }
      ]
    ];
    await context.sync();
  });
}
```

The number formatting is considered the default format. If the user, or other code, applies formatting to a cell containing a formatted number, the applied format will override the number’s format.

> ## Card layout

Reviewer 3: Broad question: 
It looks like the next three sections are very similar to sections in the existing [Entity value cards](https://learn.microsoft.com/en-us/office/dev/add-ins/excel/excel-data-types-entity-card) article. Is the main difference that these new sections describe how to add properties to CellValueTypes other than `entity`? If so, is separating the articles into entity/not-entity like this, the most logical? Should we have an article for each type? Such as one article each for entity, double, string, etc.

Cell values with properties have a default data type card that the user can view. You can provide a custom card layout to use instead of the default card layout to provide a better user experience for viewing the properties. To do this, add the **layouts** property to the JSON description.

> The **layouts** property contains two direct subproperties, **compact** and **card**. The **card** property specifies the appearance of a card when the data type card is open. The compact property is optional, and defines the icon for a value. The icon is shown in the cell value if it is provided. It can also be shown in the card if it is referenced as a sub property. See the [EntityCompactLayoutIcons](/javascript/api/excel/excel.entitycompactlayouticons) enum for a full list of available icons.

Reviewer: ```suggestion
The **layouts** property contains two direct subproperties, **compact** and **card**. The **card** property specifies the appearance of a card when the data type card is open. The compact property is optional and defines the icon for a value. The icon is shown in the cell value if it's provided. It can also be shown in the card if it's referenced as a subproperty. See the [EntityCompactLayoutIcons](/javascript/api/excel/excel.entitycompactlayouticons) enum for a full list of available icons.
```

            }
          },
          basicValue: 0.7,
          numberFormat: "00%",
          properties: {
            Computer: {
              type: Excel.CellValueType.entity,
              text: "Laptop",
              properties: {
                "Power Consumption": {
                  type: Excel.CellValueType.double,
                  basicType: Excel.RangeValueType.double,
                  basicValue: 0.25,
                  numberFormat: "00%",
                  layouts: {
                    compact: {
                      icon: "Power"
                    }
                  },
                  properties: {
                    plan: {
                      type: Excel.CellValueType.string,
                      basicType: Excel.RangeValueType.string,
                      basicValue: "Balanced"
                    }
                  }
                },
                Charging: {
                  type: Excel.CellValueType.boolean,
                  basicType: Excel.RangeValueType.boolean,
                  basicValue: true
                }
              }
            }
          }
        }
      ]
    ];
    await context.sync();
  });
}
```

> The following image shows the number value and the data type card for the nested Laptop entity.

Reviewer: ```suggestion
The following image shows the number value and the data type card for the nested laptop entity.
```

:::image type="content" source="../images/data-type-basic-nested-entities.png" alt-text="Cell value in Excel showing battery charge at 70% and the data type card showing the nested laptop entity with charging and power consuption property values.":::

> ## Attribution

Reviewer 3: For this Attribution content -- maybe we create a new article called **Data types attribution** and describe how to do attributions for all the types? I think the similarity with this [Card data attribution](https://learn.microsoft.com/en-us/office/dev/add-ins/excel/excel-data-types-entity-card#card-data-attribution) section could be confusing

> You can add attribution for information that comes from 3rd parties to indicate the source and any license information. Use **Excel.CellValueAttributionAttributes** to add attribution to a cell value. The following code example shows how to add attribution for usage of information about Mars from Wikipedia.

Reviewer: ```suggestion
Add attribution for information that comes from 3rd parties to indicate the source and any license information. Use **Excel.CellValueAttributionAttributes** to add attribution to a cell value. The following code example shows how to add attribution for usage of information about Mars from Wikipedia.
```

    const sheet = context.workbook.worksheets.getActiveWorksheet();
    const range = sheet.getRange("A1");

    range.valuesAsJson = [
      [
        {
          type: Excel.CellValueType.string,
          basicType: Excel.RangeValueType.string,
          basicValue: "cell function - Microsoft support",
          properties: {
            "Search Keywords": {
              type: Excel.CellValueType.string,
              basicType: Excel.RangeValueType.string,
              basicValue: "Cell Values"
            }
          },
          provider: {
            description: "Contoso generic search",
            //Ignacio javier igjav, Public domain, via Wikimedia Commons
            logoSourceAddress: "https://upload.wikimedia.org/wikipedia/commons/f/f9/Lupa.png",
            logoTargetAddress: "https://contoso.com"
          }
        }
      ]
    ];

    await context.sync();
  });
}
```

The following image shows how the provider information appears as the logo in the data type card for the user.

:::image type="content" source="../images/data-type-basic-provider-information.png" alt-text="Data type card showing the search logo for Contoso generic search.":::

> ## Compatibility

Reviewer 3: Should we put this section on the [Overview](https://learn.microsoft.com/en-us/office/dev/add-ins/excel/excel-data-types-overview) article? It seems relevant to all the data types

> If a user is using an older version of Excel that does not support the data types feature, they will see a warning of **Unavailable Data Type**. The value will still show in the cell and function as expected with formulas and other Excel features. If the value is a formatted number, calculations use the **basicValue** in place of the formatted number.

Reviewer: ```suggestion
On previous versions of Excel that do not support the data types feature, users will see a warning of **Unavailable Data Type**. The value will still show in the cell and function as expected with formulas and other Excel features. If the value is a formatted number, calculations use the **basicValue** in place of the formatted number.
```
